microcosm_caching/decorators.py

```diff
@@ -0,0 +1,112 @@
+from functools import wraps
+from logging import Logger
+from time import perf_counter
+from typing import Any, Type
+
+from marshmallow import Schema
+from microcosm.errors import NotBoundError
+from pymemcache.exceptions import MemcacheError
+
+from microcosm_caching.base import CacheBase
+
+
+DEFAULT_TTL = 60 * 60  # Cache for an hour by default
+
+
+def get_metrics(graph):
+    try:
+        return graph.metrics
+    except NotBoundError:
+        return None
+
+
+def cache_key(cache_prefix, key):
+    return f"{cache_prefix}:{key}"
+
+
+def cached(component, schema: Type[Schema], cache_prefix: str, ttl: int = DEFAULT_TTL):
+    """
+    Intended mainly for use as a decorator around microcosm_flask.CRUDStoreAdapters, namely
+    those revolving around retrieve functions, such that the resource is identifiable
+    by a single ID.
+
+    Decorated components must expose:
+      * an identifier_key function
+      * a logger reference
+      * the graph
+
+    :param component: A microcosm-based controller component
+    :param schema: The schema corresponding to the response type of the component
+    :param cache_prefix: Namespace to use for cache keys
+    :param ttl: How long to cache the underlying resource
+    :return: the resource (i.e. loaded schema instance)
+    """
+    identifier_key: str = getattr(component, "identifier_key")
+    logger: Logger = getattr(component, "logger")
+
+    graph = component.graph
```

Comment:
This feels like a big assumption actually. That whatever we are adding caching to will have a copy of the graph to piggyback off of.

Reply:
It's the case for the CRUDStoreAdapter, which is the main use case. It's the same assumption a bunch of other decorators have.